svc/oauth/AuthZeroOauthService.ts

```diff
@@ -0,0 +1,174 @@
+ *
+ * Oauth is supported via an integration with Auth0, a cross-platform service XH uses to provide
+ * access to Toolbox via Google, GitHub, or Microsoft identities *or* via a username/password combo
+ * created by the user in the Auth0 registration flow (and stored on Auth0 servers).
+ *
+ * This service is initialized in the `preAuthInitAsync()` lifecycle methods of the relevant
+ * AppModels. It is instantiated and initialized prior to any attempts to authenticate to the app's
+ * own Grails server. If the user is not authenticated, the Auth0 library will redirect to an Oauth
+ * flow, then return here and process the completed authentication, supplying an identity token in
+ * JWT format. This service then installs that token as a default header on all calls to the Toolbox
+ * server, which looks for the token, validates its signature to verify, and uses it to lookup or
+ * create an application user within `AuthenticationService.groovy`.
+ *
+ * TODO - preserve an incoming route for a non-authenticated user. Currently any route will be
+ *      lost during the redirect flow. We should be able to note and restore via the `state`
+ *      key we can set and then read on our Auth0 login request / post-redirect response.
+ */
+
+interface OauthConfig {
+    /** Hoist: Is OAuth enabled in this application? */
+    enabled: boolean;
+
+    /** Client ID of your app registered with Auth0 */
+    clientId: string;
+
+    /** Domain of your app registered with Auth0 */
+    domain: string;
+}
+
+export class AuthZeroOauthService extends HoistService {
+    static instance: AuthZeroOauthService;
+
+    /**
+     * Is OAuth enabled in this application?  For bootstrapping, troubleshooting
+     * and mobile development, we allow running in a non-SSO mode.
+     */
+    enabled: boolean;
+
+    auth0: Auth0Client;
+    /** Authenticated user info as provided by Auth0. */
+    user: PlainObject;
+    /** ID Token in JWT format - for passing to Hoist server. */
+    idToken: string;
+
+    /** Soft-config loaded from whitelisted endpoint on UI server. */
+    config: OauthConfig;
+
+    override async initAsync() {
+        // This service is initialized prior to Hoist auth/init, so we do *not* have our standard
+        // XH.configService ready to go at the point we need these configs. This endpoint is
+        // whitelisted in AuthenticationService.groovy to allow us to call it prior to auth.
+        const config = (this.config = (await XH.fetchJson({
+            url: 'oauthConfig'
+        }).catchDefault()) as OauthConfig);
+
+        this.enabled = config?.enabled;
+        if (!this.enabled) {
+            XH.appSpec.isSSO = false;
+            return;
+        }
+
+        if (!config?.domain || !config?.clientId) {
+            throw XH.exception(`
+                Unable to init AuthZeroOauthService - expected config not returned by server.
+                Please review the settings in configs "auth0ClientId" and "auth0Domain".
```

Comment:
I wonder if we should standardize with one config across Azure & Auth0:
`oauthConfig` JSON map
rather than these 2 separate config strings.